Remove local rewrite flag and enable it by default

alanechang opened the pull request:
This PR removes the `--rewrite-old-style-jane-street-local-annotations` flag and enables rewriting with the new syntax by default.

 A few consequences of this change are:

- `--erase-jane-syntax` now erases the old style local annotation too. See [local-erased.ml.ref](https://github.com/janestreet/ocamlformat/compare/jane...alanechang:ocamlformat:aec/remove-local-rewrite-flag#diff-be3014b26310b6710d06eb712bceab86a2417cdca4c2b5837a9ca6918a8742cc) for differences
- ~~In order for the AST equality checks to pass when we are modifying the tree deliberately, a new `ignore_local_annot_differences` option is added to ignore changes around local annotations in the standard AST. This option is set automatically when the legacy syntax is encountered. The alternative here is to convert old local syntax to the new or vice versa during normalization, which is non-trivial and will involve duplicating logic from jane syntax.~~ The standard AST normalization is changed to convert old local syntax into the new syntax in order for the equality check to pass.

alanechang committed:
remove rewrite flag

Signed-off-by: alanechang <alanechang@janestreet.com>

alanechang committed:
track local rewrites and fix ast equality check

Signed-off-by: alanechang <alanechang@janestreet.com>

alanechang committed:
rename option

Signed-off-by: alanechang <alanechang@janestreet.com>

alanechang committed:
convert old syntax to new in standard ast

Signed-off-by: alanechang <alanechang@janestreet.com>

alanechang committed:
remove local_rewrite_occurred

Signed-off-by: alanechang <alanechang@janestreet.com>

alanechang committed:
add comment

Signed-off-by: alanechang <alanechang@janestreet.com>

goldfirere commented on lib/Conf.ml:
```diff
@@ -1590,12 +1571,10 @@ let parse_state_attr attr =
-  if config.opr_opts.rewrite_old_style_jane_street_local_annotations.v then
-    String.equal test name || String.equal test ("ocaml." ^ name)
-  else false
+let is_jane_street_local_annotation _config name ~test =
```
Remove the `_config` argument. Propagating this change to callsites will unlock lots of simplification and should revert most of #38.

I will wait to review until after that change.

(I wasn't sure that this function belongs in `Conf`, but I don't actually see a better spot. Maybe `Syntax`? But that file is so small; it seems like the wrong place, too. So I say keep it here. But if you find a better spot, go for it.)

alanechang replied:
Added a commit to remove the config argument.

(I kept `is_jane_street_local_annotation` in `Conf` and couldn't find a better spot for it)

alanechang committed:
remove the unused config param

Signed-off-by: alanechang <alanechang@janestreet.com>

alanechang committed:
remove the ability to erase legacy local annots

Signed-off-by: alanechang <alanechang@janestreet.com>

alanechang commented:
Removed the logic for erasing legacy local annotation syntax in `Fmt_ast.ml` and added a note about it in the test:

https://github.com/janestreet/ocamlformat/blob/eceb6382b70e889ef85a86d128caaebd62c756f0/test/passing/tests/local_erased.ml#L1-L5

goldfirere commented on lib/Normalize_std_ast.ml:
```diff
@@ -12,6 +12,46 @@
+    (Location.mkloc name Location.none)
+    (PStr [])
+
+let convert_legacy_jane_street_local_annotations ~ghost ?segment =
```
Please add a comment explaining what transformation this is applying. I can piece it together -- and the tests show that it's right -- but a little example would go a long way to making this more maintainable.

alanechang replied:
I added a comment here together with a small example of the transformation.

goldfirere commented on lib/Normalize_std_ast.ml:
```diff
@@ -12,6 +12,46 @@
+          else [attr]
+      | _ -> [attr] )
+
+let extract_legacy_jane_street_local_annotations :
```
I'm a little worried about this function. I think it will misbehave on things like `(t[@abc][@local][@def]) -> t`, which should transform to .... _syntax error in my head_.

The problem here is that we can't have something like `(local_ (t[@abc]))[@def] -> t`. That outer `[@def]` just doesn't make sense there. That's because the `local_` is really properly understood as a modification to the `->`, not the `t`. So interjecting a `[@def]` doesn't make sense. This suggests that `[@local]` should always be the _last_ in a list of attributes. If it's not, we're in trouble and should bail out.

What we don't want is to silently change around users' code. Of course, ocamlformat checks to make sure we're not doing that.... but this code here is the very code for the test. So it pays to be careful. I'd rather error than do the wrong thing. If trouble arises in practice, we'll return to the problem case and see what to do.

alanechang replied:
I fail see the exact issue here.

`List.partition_tf` returns two list, one for things that evaluates to true and one for things that evaluates to false (in this case, it separates out the local attributes from the rest).

Something like `type t = (string[@aaa][@local][@bbb][@ccc]) -> (string[@local])` is currently valid code accepted by the compiler, and ocamlformat correctly transforms it to `type t = local_ (string[@aaa] [@bbb] [@ccc]) -> local_ string` (added as a testcase).

And the `extract_legacy_jane_street_local_annotations` function is only used in one place to move the local attributes from label_declaration into its `pld_type` to align the behavior with the new syntax.

If we are really concerned about attribute ordering, it will have to be a larger change since the ast normalization logic currently sorts the attributes with a call to `sort_attributes` and ignores reordering.

goldfirere replied:
Yes, it's fine. I didn't notice that this function is used only on label declarations. Sorry for the noise.

goldfirere commented on lib/Normalize_std_ast.ml:
```diff
@@ -132,10 +172,31 @@ let make_mapper conf ~ignore_doc_comments ~erase_jane_syntax =
+        m.expr m
+          { sbody with
+            pexp_attributes=
+              convert_legacy_jane_street_local_annotations ~ghost:is_fun
```
Is there any commentary to provide about why `~ghost:is_fun` is the right behavior here? I suppose it's OK if there isn't (this stuff _is_ a bit of a mess), but adding some insight here if you have it would be nice.

alanechang replied:
This condition turns out to not be fully correct. The only case where we want location ghost here is if **both** `is_fun` is true and it's a value binding inside of a let. Something like:

```
let g = [%local] (fun a b c -> 1)
```

The reason I imagine is because it gets rewritten to:

```
let local_ g a b c = 1
```

where the `fun` keyword is no longer present.

I pushed a new commit to fix this which will handle cases like `let g = f ([%local] (fun a b c -> 1))` correctly.

goldfirere commented on lib/Normalize_std_ast.ml:
```diff
@@ -196,6 +263,40 @@ let make_mapper conf ~ignore_doc_comments ~erase_jane_syntax =
+          { ld.pld_type with
+            ptyp_attributes=
+              convert_legacy_jane_street_local_annotations ~ghost:false
+                ~segment:"constructor_argument"
```
Is that correct? This is `label_declaration` here.

alanechang replied:
This matches the behavior of the new syntax parser which uses "constructor_argument" here and I double checked that the testsuite covers this case (changing `"constructor_argument"` here to anything else results in an error)

alanechang committed:
fix an issue around location ghost

Signed-off-by: alanechang <alanechang@janestreet.com>

alanechang committed:
add comment

Signed-off-by: alanechang <alanechang@janestreet.com>

alanechang committed:
add additional test

Signed-off-by: alanechang <alanechang@janestreet.com>

alanechang committed:
update man page

Signed-off-by: alanechang <alanechang@janestreet.com>

alanechang committed:
fix ast normalize for pattern

Signed-off-by: alanechang <alanechang@janestreet.com>

alanechang commented:
@goldfirere  Added one more commit to handle the pattern case differently. This fixes an issue caught by the CI tests.

PR should be ready to merge if that commit looks good.

goldfirere commented:
Took a look at that last commit. Looks fine. Let's merge.
